src/vsr/replica.zig

```diff
@@ -4899,112 +4948,322 @@ pub fn ReplicaType(
+            // headers from previous views. Which means 6,7 are from the current view. But since
+            // replica 0 doesn't have 6/7, then replica 1/2 must share the latest view_normal. ∎)
+
+            // Now that the canonical headers are all in place, repair any other headers:
+            for (self.do_view_change_from_all_replicas) |received| {
+                if (received) |message| {
+                    const view_normal = @intCast(u32, message.header.timestamp);
+                    assert(view_normal < message.header.view);
+
+                    if (view_normal == view_normal_canonical) continue;
+                    assert(view_normal < view_normal_canonical);
+
+                    for (message_body_as_headers(message)) |*header| {
+                        // We must trust headers that other replicas have committed, because
+                        // repair_header() will not repair a header if the hash chain has a gap.
+                        if (header.op <= message.header.commit) {
+                            log.debug(
+                                "{}: on_do_view_change: committed: replica={} op={} checksum={}",
+                                .{
+                                    self.replica,
+                                    message.header.replica,
+                                    header.op,
+                                    header.checksum,
+                                },
+                            );
+                            self.replace_header(header);
+                        } else {
+                            _ = self.repair_header(header);
+                        }
+                    }
+                }
+            }
+
+            // The `prepare_timestamp` prevents a primary's own clock from running backwards.
+            // Therefore, `prepare_timestamp`:
+            // 1. is advanced if behind the cluster, but never reset if ahead of the cluster, i.e.
+            // 2. may not always reflect the timestamp of the latest prepared op, and
+            // 3. should be advanced before discarding the timestamps of any uncommitted headers.
+            const head = self.journal.header_with_op(self.op).?;
+            if (self.state_machine.prepare_timestamp < head.timestamp) {
+                self.state_machine.prepare_timestamp = head.timestamp;
+            }
+
+            // Any uncanonical ops remaining either:
+            // * Connect to the hash chain on the right.
+            // * Do not connect on the right (hash chain break).
+            //
+            // If there is a hash chain break, none of the headers from the canonical DVCs replaced
+            // the broken op. It is truncated like a gap.
+            //
+            // Removing these is necessary for correctness and liveness, to ensure that
+            // disconnected headers do not remain in place in lieu of gaps.
+            var op: u64 = op_uncanonical - 1;
```

Comment:
Is it possible where a replica is so far behind that its uncanonical op is actually a few tens of ops behind, also completely disconnected, then we remove everything to the right, and then participate in a subsequent view change, but having lost what were committed headers (what we removed)?

On second thought, I guess not, since it's bounded by `pipeline_max` and we've noted this also in comments. But I thought I'd ask just to double check.

Reply:
Right, it is bounded by `pipeline_max`. Also if the header is possibly committed, it can't be disconnected at this point. The other DVCs would have provided the headers to repair the committed header's hash chain, due to the quorum intersection property.